src/lib/SaferNames/Type.hs

```diff
@@ -0,0 +1,968 @@
+-- Copyright 2019 Google LLC
+--
+-- Use of this source code is governed by a BSD-style
+-- license that can be found in the LICENSE file or at
+-- https://developers.google.com/open-source/licenses/bsd
+
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE StandaloneDeriving #-}
+
+module SaferNames.Type (
+  checkModule, checkTypes, getType, litType, getBaseMonoidType) where
+
+import Prelude hiding (id)
+import Control.Category ((>>>), id)
+import Control.Monad
+import Control.Monad.Reader
+import Control.Monad.Except hiding (Except)
+import Control.Monad.Identity
+import Data.Foldable (toList)
+import Data.Functor
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Map.Strict as M
+import qualified Data.Set        as S
+import Data.Text.Prettyprint.Doc
+
+import LabeledItems
+
+import Err
+import Type (litType)
+import Util (forMZipped, forMZipped_, iota)
+
+import SaferNames.Syntax
+import SaferNames.Name
+import SaferNames.PPrint ()
+
+-- === top-level API ===
+
+checkModule :: (Distinct n, MonadErr m) => TopBindings n -> Module n -> m ()
+checkModule topBindings m =
+  addContext ("Checking module:\n" ++ pprint m) $ asCompilerErr $
+    runBindingsReaderT (fromTopBindings topBindings) $
+      checkTypes m
+
+checkTypes :: (BindingsReader m, MonadErr1 m, CheckableE e)
+           => e n -> m n ()
+checkTypes e = do
+  Distinct scope <- getScope
+  bindings <- getBindings
+  liftEither $ runTyperT (scope, bindings) $ void $ checkE e
+
+getType :: (BindingsReader m, HasType e)
+           => e n -> m n (Type n)
+getType e = do
+  Distinct scope <- getScope
+  bindings <- getBindings
+  return $ runIgnoreChecks $ runTyperT (scope, bindings) $ getTypeE e
+
+-- === the type checking/querying monad ===
+
+-- TODO: not clear why we need the explicit `Monad2` here since it should
+-- already be a superclass, transitively, through both MonadErr2 and
+-- MonadAtomSubst.
+class ( MonadFail2 m, Monad2 m, MonadErr2 m, EnvGetter Name m
+      , BindingsReader2 m, BindingsExtender2 m)
+     => Typer (m::MonadKind2) where
+  declareEffs :: EffectRow o -> m i o ()
+  extendAllowedEffect :: Effect o -> m i o () -> m i o ()
+  withAllowedEff :: EffectRow o -> m i o a -> m i o a
+
+-- This fakes MonadErr by just throwing a hard error using `error`. We use it
+-- to skip the checks (via laziness) when we just querying types.
```

Comment:
Interesting! Have you verified that it actually skips the checks?

Reply:
No, I should try using Debug.Trace to make sure